lib/nostrum/util.ex

```diff
@@ -204,6 +204,39 @@ defmodule Nostrum.Util do
       String.to_atom(token)
   end
 
+   # Generic casting function
+  @doc false
+  @spec cast(term, module | {:list, term} | {:struct, term} | {:index, term, [atom]}) :: term
+  def cast(value, type)
+  def cast(nil, {:list, _}), do: []
+  def cast(nil, _type), do: nil
+
+  def cast(values, {:list, type}) when is_list(values) do
+    Enum.map(values, fn value ->
+      cast(value, type)
+    end)
+  end
+
+  # This takes `value` list and casts it into a map of type indexed by `index_by`.
+  def cast(value, {:index, type, index_by}) when is_list(value) do
+    new_list = cast(value, {:list, type})
+
+    Map.new(new_list, fn item ->
+      {struct_get_in(item, index_by), item}
+    end)
+  end
+
+  def cast(value, {:struct, module}) when is_map(value) do
+    module.to_struct(value)
+  end
+
+  def cast(value, module) do
```

Comment:
This really only applies to the `Snowflake` module currently, correct? Do you see a use case in the future for this? Probably if we need to implement another "custom" type?

Reply:
I did see a potential case for keeping this in the `Nostrum.Struct.Guild` module. That module has indexed data fields (:members, :channels, :roles) that could each be represented as an individual type. 

We can remove that code for sure, but I think it would be good to open up that conversion function for extension.

Reply:
I see what you mean, we can leave it her for now then.